1단계 - 레거시 코드 리팩터링

ghtjr410 opened the pull request:
# 레거시 코드 리팩터링 과제 제출
안녕하세요!
레거시 코드 리팩터링 과제 제출합니다.
## 미션 진행 과정
1. **TDD Step 설계**: 리팩터링 전 단계별 작업 리스트업
2. **Bottom-Up 방식**: Answer → Answers(일급 컬렉션) → Question 순서로 진행
3. **작은 단위로 분리 후 조합**: Leaf 모듈부터 분리하고 Composite하는 방식
## 고민했던 지점

### 1. 예외를 어디서 던질 것인가?
- Answer가 직접 `validateOwner()` 예외를 던지면 특정 유스케이스(삭제)에 종속됨
- **결론**: Answer는 `isOwner()`로 사실만 응답, Question이 삭제 맥락에서 예외 처리

### 2. 삭제와 이력 생성 분리
- 기존: `delete()` 호출 시 상태 변경 + 이력 반환을 한 번에
- **개선**: `delete()`는 상태 변경만, `deleteHistories()`는 이력 생성만 (SRP)

### 3. 시간 파라미터 주입
- `LocalDateTime.now()` 내부 호출 시 테스트 어려움
- **해결**: Service에서 시간 생성 후 파라미터로 주입
- **향후 개선**: `ClockHolder` 같은 인터페이스로 추상화하면 Service 레이어 테스트도 용이해짐

## 의식적인 연습

### 레거시 리팩터링 전략
- **Leaf 우선 분리**: 가장 쉽게 분리 가능한 것부터 찾아서 분리
- **작은 사이클 반복**: 한 번에 뭉텅이로 분리하면 사이드이펙트 발생 위험
- **단계별 테스트**: 분리할 때마다 테스트로 검증
- **빠른 반복**: 작은 리팩터링을 빠르게 반복하면 큰 리팩터링보다 안정적이고 비슷하거나 더 빠름

---

제가 실무에서 개선 중인 레거시들은 여러 도메인이 복잡하게 얽혀있어서 
제공해주신 레거시코드는 처음보자마자 머리속에 어떻게 리팩터링해야할지 설계가 되더라구요. 
그래서 바로 단계별 작업 리스트업하고 진행하였더니 너무 재밌게 즐기면서 과제 진행했습니다.

리뷰 잘 부탁드립니다!

ghtjr410 committed:
feat: 개발 환경 자동화 및 코드 포맷터(Spotless) 적용

.gitignore
- diff 디렉토리 무시 규칙 추가

.sdkmanrc
- Java 버전 설정 파일 추가 (11.0.21-tem)

Makefile
- diff 생성/정리, validation, init 등 스크립트 실행 타겟 추가

build.gradle
- Spotless 플러그인 추가
- Java 포맷팅 규칙(palantir format 등) 설정 적용

clean-diff.sh
- diff/commit 디렉토리 내 패치 파일 정리 로직 추가

commit-diff.sh
- 스테이지된 변경사항을 timestamp 포함 패치 파일로 저장 기능 추가

origin-ghtjr410-diff.sh
- origin/ghtjr410 대비 현재 브랜치 diff 생성 스크립트 추가

pre-commit
- Java 변경 시 Spotless 검사 자동 수행 훅 추가

init.sh
- Git hook 설정 및 스크립트 실행 권한 부여 자동화

validate.sh
- Spotless 적용/검사 및 프로젝트 빌드를 자동 수행하는 검증 스크립트 추가

ghtjr410 committed:
docs: README 수정 및 1단계 리팩터링 문서/체크리스트 추가

README.md
- 프로젝트 제목 수정 및 LMS/TDD 소개 추가
- 단계별 문서 링크(1~4단계) 추가

01-legacy-code-refactoring.md
- 1단계 리팩터링 학습 목표, 요구사항, 힌트, PR 체크 항목 문서화

checklist.md
- 미션 제출 전 객체지향 원칙, TDD 원칙, 테스트 품질 등 체크리스트 추가

ghtjr410 committed:
style: Spotless 코드 포맷팅 적용

ghtjr410 committed:
refactor: Answer 삭제 로직을 도메인 메서드로 캡슐화하고 테스트 추가

Answer.java
- setDeleted(boolean) 메서드 제거
- delete(LocalDateTime) 도메인 메서드 추가하여 삭제 처리 및 DeleteHistory 생성 일원화
- delete 호출 시 deleted 플래그 true 설정

QnAService.java
- Answer 삭제 시 setDeleted + DeleteHistory 생성 로직을 answer.delete(now) 호출로 단순화

AnswerTest.java
- delete 메서드 동작 검증 테스트 추가
- 삭제 후 DeleteHistory 반환 및 deleted 플래그 true 검증

ghtjr410 committed:
refactor: Question 삭제 로직 도메인으로 이동 및 권한 검증 테스트 추가

Question.java
- isOwner 메서드를 private으로 변경하여 외부 직접 호출 차단
- setDeleted(boolean) 제거
- delete(loginUser, now) 메서드 추가하여 권한 검증 및 DeleteHistory 생성 일원화
- 삭제 권한 없을 경우 CannotDeleteException 발생 처리 추가

QnAService.java
- 서비스 레이어에서의 직접 권한 검사 및 삭제 처리 로직 제거
- question.delete(...) 호출로 삭제 처리 단일화

QuestionTest.java
- 질문 삭제 성공 테스트 추가
- 소유자 불일치 시 CannotDeleteException 발생 테스트 추가

ghtjr410 committed:
refactor: 질문·답변 삭제 로직을 Question 도메인으로 완전 통합하고 검증 강화

Question.java
- getAnswers 제거하여 외부 접근 최소화
- delete 메서드를 Question 내부에서 답변까지 함께 삭제하도록 변경
- validateDeletable 추가: 질문 소유자 검증, 타인 답변 존재 시 예외 처리
- delete 시 질문/답변 DeleteHistory 목록 생성 후 반환

QnAService.java
- 서비스 레이어 삭제 검증 및 반복 로직 제거
- question.delete(...) 호출만으로 삭제 처리 일원화

QuestionTest.java
- 삭제 성공 테스트 수정: 답변 존재 시 DeleteHistory 리스트 검증
- 소유자 불일치 예외 검증 추가
- 타인 답변 존재 시 CannotDeleteException 발생 테스트 추가

ghtjr410 committed:
refactor: 중복 삭제 메서드 및 불필요한 접근자 정리

Answer.java
- delete(LocalDateTime) 메서드를 클래스 상단으로 이동하고 중복 정의 제거
- 불필요한 getContents() 제거

Question.java
- delete 메서드 및 validateDeletable 메서드를 상단으로 재배치하고 중복 블록 제거
- title/contents setter 제거로 불변성 강화
- 전체 구조 정리하여 중복 코드 제거 및 가독성 향상

ghtjr410 committed:
refactor: Answer 목록 전용 값 객체 Answers 추가 및 삭제/검증 로직 위임

Answers.java
- Answer 리스트를 캡슐화하는 값 객체 추가
- deleteAll(now): 모든 Answer 삭제 수행 후 DeleteHistory 리스트 반환
- validateDeletable(loginUser): 삭제 가능 여부 검증(타인 답변 존재 시 예외 발생)

AnswersTest.java
- 소유자 일치 시 예외 없이 통과하는 검증 테스트 추가
- 소유자 불일치 시 CannotDeleteException 발생 테스트 추가

ghtjr410 committed:
refactor: Question이 Answer 컬렉션을 Answers 값 객체로 위임하도록 구조 개선

Answers.java
- add 메서드 추가로 Answer 등록 책임 부여
- deleteAll(now) 기존 로직 유지
- validateDeletable(loginUser) 그대로 활용

Question.java
- answers 필드를 List → Answers 로 교체하여 컬렉션 캡슐화
- 삭제 시 answers.deleteAll(now) 사용으로 책임 위임
- 검증 로직도 answers.validateDeletable(loginUser) 호출로 단순화

ghtjr410 committed:
refactor: 공통 엔티티 속성을 BaseEntity로 분리하고 Answer·Question에 상속 적용

Answer.java
- BaseEntity 상속 적용으로 id/createdDate/updatedDate 필드 제거
- 생성자에서 super(id) 호출로 id 설정 위임
- delete 시 getId() 사용하도록 변경
- 중복 필드 제거로 엔티티 구조 단순화

BaseEntity.java
- id, createdDate, updatedDate 공통 필드 추가
- getId() 제공
- Answer/Question의 공통 속성을 추출한 추상 엔티티 생성

Question.java
- BaseEntity 상속 적용 및 id 관련 필드 제거
- 생성자에서 super(id) 호출하도록 변경
- DeleteHistory 생성 시 getId() 사용하도록 수정

ghtjr410 committed:
refactor: Question의 title/contents를 전용 값 객체로 분리하여 응집도 향상

Question.java
- title, contents 필드를 제거하고 QuestionContent 값 객체로 대체
- 생성자에서 QuestionContent 생성하여 content 필드에 저장
- toString 수정: content 정보 출력하도록 변경
- 전체 구조를 값 객체 기반으로 단순화

QuestionContent.java
- title, contents를 보관하는 값 객체 추가
- toString 재정의하여 도메인 표현 명확화

ghtjr410 committed:
test: Question 삭제 테스트 간소화 및 불필요한 테스트 제거

QuestionTest.java
- 삭제 성공 테스트 수정: 반환된 DeleteHistory 리스트 크기 검증 후 QUESTION 기록 포함 여부만 확인하도록 단순화
- 삭제 여부(isDeleted) 검사 유지
- 중복되는 ‘다른 사람 답변 예외’ 테스트 제거(AnswersTest로 책임 분리됨)

ghtjr410 committed:
refactor: 연관관계 설정 위치 정리 및 QuestionContent 접근자 추가

Answer.java
- toQuestion 위치 조정 및 중복 메서드 제거
- 불필요한 코드 정리로 클래스 구조 단순화

Question.java
- addAnswer에서 Answer ↔ Question 연관관계 설정(toQuestion) 명확히 수행
- getWriter 위치 조정 및 중복 제거
- 전체 구조 정돈

QuestionContent.java
- title, contents 조회용 getter 추가
- 값 객체로서 외부 접근이 필요한 영역을 최소한으로 허용

ghtjr410 committed:
feat: Answer 삭제 이력 생성 기능 분리 추가

Answer.java
- deleteHistory(LocalDateTime) 메서드 추가하여 삭제 여부 변경 없이 DeleteHistory만 생성하도록 분리

AnswerTest.java
- deleteHistory 동작 검증 테스트 추가
- 특정 시각을 지정해 DeleteHistory 생성 값 검증

ghtjr410 committed:
feat: 삭제 이력 생성 기능을 엔티티/컬렉션에서 분리 제공하도록 확장

Answers.java
- deleteHistories(deletedAt) 추가: 삭제 플래그 변경 없이 Answer들의 DeleteHistory만 생성
- 책임 분리로 deleteAll vs deleteHistories 용도 명확화

Question.java
- deleteHistories(deletedAt) 추가: Question + Answers의 DeleteHistory 목록 생성 전용 메서드 제공
- 삭제 여부 변경 없이 이력 생성만 수행

QuestionTest.java
- deleteHistories 테스트 추가
- Question 및 Answer의 DeleteHistory 생성 여부 검증

ghtjr410 committed:
refactor: 삭제 행위와 삭제 이력 생성을 완전히 분리하여 도메인 책임 명확화

Answer.java
- delete(LocalDateTime) → delete()로 변경하여 삭제 상태만 변경
- 삭제 이력 생성을 담당하는 deleteHistory(LocalDateTime) 분리
- 내부 delete 로직 단순화 및 책임 분리

Answers.java
- deleteAll(now) 제거 → deleteAll()로 변경하여 상태 변경만 수행
- 삭제 이력 생성 기능은 deleteHistories(LocalDateTime)로 분리하여 순수 히스토리 생성 책임만 담당

Question.java
- delete(loginUser, now) → delete(loginUser)로 변경해 삭제 시각 의존 제거
- 내부적으로 Answer 삭제는 answers.deleteAll() 호출로 단순화
- 삭제 이력 생성은 deleteHistories(LocalDateTime)에서 전담

QnAService.java
- 도메인 delete 호출 후 별도로 deleteHistories(now) 호출하여 이력 생성
- 삭제와 삭제 이력 저장을 명확히 분리

AnswerTest.java
- delete()는 상태 변경만 검증하도록 테스트 단순화

QuestionTest.java
- delete() 결과로 DeleteHistory가 반환되지 않으므로 관련 검증 제거
- 삭제 여부 검증만 유지
- 소유자 불일치 테스트는 delete(loginUser)로 수정

ghtjr410 committed:
refactor: 삭제 시각을 고정하여 일관된 DeleteHistory 생성

QnAService.java
- LocalDateTime.now()를 한 번만 호출해 fixedNow로 저장
- question.delete(loginUser) 이후 deleteHistories 생성 시 fixedNow를 재사용하여 삭제 이력 시간 불일치 문제 해결

ghtjr410 committed:
docs: 리팩터링 PR 링크 및 구현 기능 목록 문서에 추가

01-legacy-code-refactoring.md
- PR 링크 추가
- 구현 기능 목록 상세 추가(도메인 리팩터링·서비스 로직·테스트 항목 정리)
- 전체 리팩터링 진행 내역을 문서로 구조화하여 가독성 향상

javajigi commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -43,13 +35,16 @@ public Answer(Long id, NsUser writer, Question question, String contents) {
-        return this;
+    public void delete() {
+        this.deleted = true;
+    }
```
현재 구현한 방식과 아래와 같이 글쓴이 인지 여부를 이 메서드가 판단하도록 구현하는 것 중 어느 방식으로 구현하는 것이 좋을지 고민해 본다.
위 두 가지 방식 중 선택한 방법으로 구현하는 것이 좋은 이유에 대해 피드백으로 남겨본다.

```java
    public void delete(NsUser loginUser){
        // 글쓴이 인지를 판단하는 로직 검증 후 삭제 상태로 변경
    }
```

이 부분은 RacingGame을 다음과 같이 구현한 것과 비슷한 관점으로도 볼 수 있지 않을까?

```java
    public void move(int randomValue) {
        validateRandomValue(randomValue);

        if (shouldMoveForward(randomValue)) {
            this.distance += MOVE_DISTANCE;
        }
    }
```

javajigi commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -43,13 +35,16 @@ public Answer(Long id, NsUser writer, Question question, String contents) {
+        this.deleted = true;
+    }
+
+    public DeleteHistory deleteHistory(LocalDateTime deletedDate) {
```
삭제와 DeleteHistory 생성을 분리한 점 👍

javajigi commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,45 @@
+import nextstep.users.domain.NsUser;
+
+public class Answers {
+    private final List<Answer> values;
```
일급 콜렉션 적용 👍

javajigi commented on src/main/java/nextstep/qna/domain/BaseEntity.java:
```diff
@@ -0,0 +1,19 @@
+
+import java.time.LocalDateTime;
+
+public abstract class BaseEntity {
```
공통의 값에 대해 부모 객체 분리 👍

javajigi commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -1,92 +1,77 @@
-
-    private String contents;
-
+public class Question extends BaseEntity {
```
```suggestion
public class Question extends DeletableBaseEntity {
```

"이 프로젝트의 정책이 모든 컨텐츠는 삭제는 Soft Delete 정책을 따른다."와 같은 정책이 있다고 상상해 본다면 아래와 같이 부모 객체를 설계하는 것도 한 방법이 될 수 있겠다.

- DeletableBaseEntity는 BaseEntity를 상속
- DeletableBaseEntity가 deleted 상태 값을 가지고 상태 변경 여부 메서드를 가진다.

javajigi commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -1,92 +1,77 @@
-    private LocalDateTime createdDate = LocalDateTime.now();
-
-    private LocalDateTime updatedDate;
+    private boolean deleted = false;
```
원시 값 포장이 의미있을까?

javajigi commented on src/main/java/nextstep/qna/domain/QuestionContent.java:
```diff
@@ -0,0 +1,24 @@
+package nextstep.qna.domain;
+
+public class QuestionContent {
```
객체 분리 👍

javajigi commented on src/main/java/nextstep/qna/service/QnAService.java:
```diff
@@ -26,24 +24,12 @@ public class QnAService {
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
+
+        LocalDateTime fixedNow = LocalDateTime.now();
+
+        question.delete(loginUser);
+        List<DeleteHistory> deleteHistories = question.deleteHistories(fixedNow);
```
👍

javajigi commented on src/test/java/nextstep/qna/domain/AnswerTest.java:
```diff
@@ -1,8 +1,35 @@
+    @Test
+    void deleteHistory() {
+        Answer answer = new Answer(NsUserTest.JAVAJIGI, QuestionTest.Q1, "Answers Contents1");
+        LocalDateTime fixedTime = LocalDateTime.of(2025, 1, 1, 10, 0);
+
+        DeleteHistory history = answer.deleteHistory(fixedTime);
```
시간에 대한 테스트까지 고려한 점 👍

javajigi reviewed (changes requested):
1단계 미션 진행 💯 
특별히 피드백할 내용이 없을 만큼 구현 잘 했네요.
단, 같이 고민해 봤으면 하는 내용이 있어 피드백 남겨봤어요.
추가로 좀 더 고민해본 후에 다음 단계로 넘어가면 좋겠습니다.

ghtjr410 committed:
refactor: 엔티티 공통 삭제 기능을 DeletableBaseEntity로 통합하고 Answer/Question 구조 개선

Answer.java
- BaseEntity → DeletableBaseEntity 상속으로 변경
- deleted 필드 제거 (부모로 이동)
- delete() 제거 및 상위 protected delete() 활용
- deleteHistory 메서드는 그대로 유지
- toString 개선 및 중복 코드 정리

BaseEntity.java
- 기존 BaseEntity 삭제 (DeletableBaseEntity로 통합)

DeletableBaseEntity.java
- id, createdDate, updatedDate, deleted 공통 관리
- delete(), isDeleted() 제공
- 엔티티 공통 toString 구현

Question.java
- BaseEntity → DeletableBaseEntity 상속 변경
- deleted 필드 제거 (부모 클래스에서 관리)
- delete() 내부에서 상위 delete() 호출
- isDeleted() 제거 (상위 클래스 제공)
- toString 개선 및 구조 단순화

ghtjr410 committed:
feat: Answer 삭제 시 소유자 검증 로직 추가 및 테스트 보강

Answer.java
- delete(NsUser) 추가: 삭제 전 소유자 검증 수행
- 소유자 불일치 시 CannotDeleteException 발생
- 기존 delete() 로직을 상위 delete() 호출로 위임

AnswerTest.java
- delete_소유자일치_성공 테스트 추가
- delete_소유자불일치_예외발생 테스트 추가
- 삭제 이력 생성 테스트(deleteHistory)는 그대로 유지

ghtjr410 committed:
refactor: Answer 개별 삭제 책임 이전으로 Answers의 삭제 검증 제거 및 관련 테스트 삭제

Answers.java
- validateDeletable 제거
  → 소유자 검증 책임을 Answer.delete(loginUser)로 완전히 이관
- Answers는 컬렉션 관리와 일괄 삭제(deleteAll)·이력 생성(deleteHistories)만 담당

Question.java
- validateDeletable 삭제에 따라 내부 검증 로직 단순화
  → 질문 소유자 검증만 수행
  → 답변 삭제는 answers.deleteAll()이 Answer.delete()를 호출하며 처리

AnswersTest.java
- validateDeletable 제거에 따라 해당 테스트 클래스 전체 삭제

ghtjr410 committed:
refactor: 삭제 권한 검증을 Answer로 위임하고 출력 포맷 및 일괄 삭제 로직 개선

Answer.java
- toString에서 question 정보 제거하여 간결화

Answers.java
- deleteAll(loginUser)로 변경하여 각 Answer가 소유자 검증을 수행하도록 위임
- Answer.delete(loginUser) 호출 구조로 일관성 확보

DeletableBaseEntity.java
- toString 포맷 수정(불필요한 '}' 제거)

Question.java
- delete 수행 시 answers.deleteAll(loginUser) 호출로 삭제 권한 검증 일관화

ghtjr410 committed:
test: Question 삭제 테스트 이름을 의도에 맞게 명확화

QuestionTest.java
- `delete_성공` → `delete_소유자일치_성공` 으로 테스트명 변경하여 검증 목적을 더 분명하게 표현

ghtjr410 committed:
refactor: 삭제 도메인 로직 및 테스트 명세 정리

01-legacy-code-refactoring.md
- Answer: 삭제 시 소유자 검증 및 상태 변경 단계 명시
- Answers: 불필요해진 삭제 가능 여부 검증 항목 제거
- Question: 삭제 처리 흐름을 단순화(답변 삭제 포함)
- DeletableBaseEntity: 공통 삭제 필드 및 delete() 기능 명세 추가
- 테스트: 메서드명과 시나리오를 명확하게 정리(성공/실패 구분)

ghtjr410 commented:
# 피드백 반영

## 변경 사항

### 1. BaseEntity → DeletableBaseEntity
- Soft Delete 정책을 명확히 드러내는 네이밍으로 변경
- `deleted` 필드를 부모 클래스로 이동하여 중복 제거

### 2. delete() 메서드에서 소유자 검증
- Answer.delete(NsUser)로 변경하여 내부에서 소유자 검증 후 삭제
- 검증과 상태 변경을 하나의 메서드에서 처리 (RacingGame 예시 참고)

### 3. AnswersTest 제거
- Answer.delete()가 소유자 검증까지 담당하므로 Answers의 validateDeletable() 불필요
- 단순 위임 로직만 남아 별도 테스트 불필요

---

이번에도 리뷰 잘 부탁드립니다!

javajigi reviewed (approved):
피드백 반영 👍 
현장의 레거시 코드 복잡도가 많이 높겠지만 이 미션과 같이 점진적인 리팩터링을 통해 레거시 코드를 리팩터링해보는 경험을 쌓아보면 좋겠습니다.
2단계가 db 없이 도메인 로직 구현부터 해보는 경험인데요.
이 미션을 통해서 새로운 접근 방식에 도전해 보면 좋겠습니다.
